stout/grpc/client.h

```diff
@@ -233,24 +257,37 @@ struct _Call {
       WriteMaybeLast(request, options, false);
     }
 
+    // Equivalent to calling `Write()` and `WritesDone()` one after the other,
+    // although possibly more efficient.
+    //
+    // You must call exactly one of `WritesDone()` or `WriteLast()` at the end
+    // of a message stream.
     void WriteLast(
         const RequestType_& request,
         const ::grpc::WriteOptions& options = ::grpc::WriteOptions()) {
       WriteMaybeLast(request, options, true);
     }
 
+    // Signals that the stream of messages coming from the client is complete.
+    //
+    // You must call exactly one of `WritesDone()` or `WriteLast()` at the end
+    // of a message stream.
     void WritesDone() {
       bool write = false;
 
       mutex_.lock();
+      assert(!finish_); // Only call one of WriteLast() or WritesDone() once.
```

Comment:
I applaud the enforcement, but I see two issues here.

1) `assert` is brutal: you can't `catch` it and thus this is not a very friendly way of failing. My suggestion is to `throw` an appropriate exception instead. Exceptions in C++ are performance-wise rather expensive though, which leads me to (2).

2) As `finish_` is `private` - and as there is no getter - any user of the class is forced to do their own book keeping of whether it is "safe" to call `WriteLast` or `WritesDone`. Alternatively a user might wrap their calls in `try`/`catch` which I would consider an anti-pattern. Exposing a getter would allow a user to determine if it is safe to call: i.e., `if(!client->isFinished()) { client->WriteLast(...); }`